tls/ttls.c

```diff
@@ -2203,15 +2204,19 @@ ttls_recv(void *tls_data, unsigned char *buf, size_t len, unsigned int *read)
 		if (io->msgtype == TTLS_MSG_HANDSHAKE
 		    && ttls_hs_checksumable(tls))
 		{
-			if (likely(*read - parsed >= TTLS_HS_HDR_LEN)) {
+			if (likely(*read - parsed != len)) {
```

Comment:
Having `*read - parsed != len` doesn't mean `ttls_parse_record_hdr()` ate whole header and there is still data in the buffer. It means the function ate at least last bytes of a handshake header and there is still payload available.

So, if we get handshake in two pieces, one of one byte and other of everything else, we'll have `*read - parsed != len`, where `*read - parsed` not equal to `TTLS_HS_HDR_LEN`. That will break the checksum calculation trick.

Reply:
No.  `ttls_parse_record_hdr()` returns `T_POSTPONE` if it has not eaten the whole header (`TLS_HEADER_SIZE + TTLS_HS_HDR_LEN`) and does not reach here

Reply:
Wait a minute, I'm confused.
Say, a single-byte packet arrives. Then `len` is 1, we call `ttls_parse_record_hdr()`, it returns `T_POSTPONE`, we return that to the called. Then, a large packet arrives. To simplify things, let's assume it contains everything else in the TLS record. We pass that to `ttls_parse_record_hdr()`. It increases `*read` by some number, but that number is not `len`, as there is a payload after a header. We go into the "then" branch, set `hh_len` to `TTLS_HS_HDR_LEN`, which subsequently will trigger the checksumming trick. We'll calculate checksum not from `buf`, but from `buf - hh_len`. But we don't have that first single byte in the continuous buffer. It was in the previous skb, not in the current one.

Is that right? Or am I missing something here?

Reply:
I understand what you mean. You're right. The trick with `hh_len` is the problem

Reply:
OK.

Also, that `TTLS_HS_HDR_LEN` looks suspicious. I mean, there is the record header, which is `TLS_HEADER_SIZE` bytes long. And then there is handshake header, that takes additional `TTLS_HS_HDR_LEN` bytes. So shouldn't there be sum of those two, that is 9 bytes instead of 5?

Reply:
Not. We are trying to fix the case when 9 bytes just arrive and they are discarded as we assign 4 (https://github.com/tempesta-tech/tempesta/blob/master/tls/ttls.c#L2211) to the variable and hope to update the checksum later, but exit here (https://github.com/tempesta-tech/tempesta/blob/master/tls/ttls.c#L2221)
But unfortunately the problem is not only in this, but in general in the presence of a variable `hh_len`

Reply:
Yep, first 9 bytes are the headers. There is no sense in postponing checksum calculation if there will be no checksum calculation on the buffer later. But if `ttls_parse_record_hdr()` returns `9`, and it's a handshake record, and there is more than 9 bytes in the buffer, then there will be further processing, and later we'll calculate checksum. Otherwise there will be no occasion, so we need to update checksum immediately. How about that?

Reply:
There was an off-site discussion, so I'll put here the bits that were figured out.

We do not checksum TLS header, only handshake header, so we need to check for 4 bytes, which is `TTLS_HS_HDR_LEN`. If we have 4 bytes parsed and there are more data in the buffer, we still can checksum both the header and the payload in a single call. The condition (pre-patch) looks fine. Although it's questionable what will happen if we parsed 4 bytes and there are no bytes left in the buffer. Or what will happen if buffer starts with remaining bits of a TLS header, and the like.